### DISCUSSION TITLE
Invoking a QLACK3 WD application from another WD application

### DISCUSSION THREAD

**Discussion by european-dynamics-rnd** [qlack/QLACK-WebDesktop#112] [Q&A]
Good afternoon QLACK team,

Is it possible to invoke a QLACK3 WD application from another WD application?

The scenario that we have is that while in an QLACK3 WD application and specifically Moodle LMS, clicking on a hyperlink, should either open another application and specifically the Alfresco Document Library in a specific path/context, or if it is already opened it should refresh the UI to show the right content/context.

I remember in the past some from the team mentioned using the qlack-pubsub.js mechanism. Is this still the case and if yes how we can use it with applications that we cannot control their source? 

As an example I have created/integrated two simple web-applications and by subscribing to the same QPubSub channel I was able to update the iFrame application-A with what application-B sent as a request.

Thank you,
-Alkis

**Comment by NMichas** [comment]
Hello, indeed, WD provides a built-in inter-application messaging solution based on `window.postMessage()`.

You can find relevant examples here:

- https://github.com/qlack/QLACK-Samples/tree/master/QLACK-WebDesktop-pubsub-toast-sample
- https://github.com/qlack/QLACK-Samples/tree/master/QLACK-WebDesktop-pubsub-user-profile-sample

**Reply by european-dynamics-rnd** [reply]
So this means that we have to modify and rebuild the source code of the solutions we have integrated. 
Unfortunately since these are the Moodle and Alfresco applications this is not be possible.

What are out alternatives? I mean is there a plan to allow this natively when deploying an application in WD, to configure the channel that we want to subscribe it, so that when it receives a specific request (you will define the type of messages that are allowed, for example: open, close, gotourl, ....) an action is performed?

**Comment by NMichas** [comment]
We don't have such plans, although even if such functionality existed you'd still have the problem at source (i.e. where you generate the link) as, for example, when you click on a link/button it would need to send a message to WD instead of its default action (so, you'd still need to modify the original source code).

If you do have a way to generate links at source (that result in calling your own JS code), you can have a third application listening for these messages and then manually manipulating the windows (IFrames) of WD to control the target application.

**Reply by european-dynamics-rnd** [reply]
Is there an example code, showing how I can manually manipulate the windows (IFrames) of WD to control a target application? 

It seems that your suggestion of a 3rd application, listening for events could work. However, does this 3rd application, need to be open in WD and visible? IS there a way to instruct WD to launch by default an WD application?

As an example I could ask the partner in charge of Moodle (Appl-A), to "change the code" and send a message on a preconfigured channel via the qlack-pubsub.js mechanism. Then the 3rd application (Appl-C) could catch this message and simply instruct (manipulate) Alfresco application (Appl-B) to navigate to the path, indicated in the body of the message. Would something like that work?

**Comment by NMichas** [comment]
> As an example I could ask the partner in charge of Moodle (Appl-A), to "change the code" and send a message on a preconfigured channel via the qlack-pubsub.js mechanism. Then the 3rd application (Appl-C) could catch this message and simply instruct (manipulate) Alfresco application (Appl-B) to navigate to the path, indicated in the body of the message. Would something like that work?

Yes, that's the idea. The application should be open (to receive the events) but can be minimised. Unfortunately, we don't have such an example.

**Reply by european-dynamics-rnd** [reply]
Thanks @NMichas once again for your quick response.
Is is possible to talk with someone from the dev team to show me how I can "manually manipulate the windows (IFrames) of WD to control a target application"?

**Comment by NMichas** [comment]
I don't think anyone tried this before, so not sure how we could help.

Another option would be to extend the default WD event listener to support custom messages, one of them being URL refreshes for a specific window. You can see how WD listens to generic events here: https://github.com/qlack/QLACK-WebDesktop/blob/0033c2b5af925eff463ea79df6f3765b9b8e187a/qlack-webdesktop-ui/angular/src/app/common/start-menu/start-menu.component.ts#L119

**Reply by european-dynamics-rnd** [reply]
How long will the option of extending the default WD event listener might take to create? 
I assume that I could refresh the URL for a specific/configurable window. Right?

**Comment by NMichas** [comment]
Opened applications are kept here:
https://github.com/qlack/QLACK-WebDesktop/blob/0033c2b5af925eff463ea79df6f3765b9b8e187a/qlack-webdesktop-ui/angular/src/app/common/start-menu/start-menu.component.ts#L19

You should get a reference to the application, find its window, and issue a new URL for the embedded IFrame. Not sure how long it might take as we never tried something like that before.
